Merge OAuth chapter into master

tghosth opened the pull request:
<!---
IMPORTANT NOTES:
- Changes should always be made only in the raw .md files and not in the CSV, JSON, XLSX, PDF, DOCX files, etc.
- Please do not open a pull request without first opening an associated issue.
- Please carry out all discussion in the associated issue only.
- Please refer to the following link for guidance on labeling contributions https://github.com/OWASP/ASVS/blob/master/CONTRIBUTING.md#standard-for-changes
-->

This Pull Request relates to issue #996.

For now, this is just to monitor the "mergability" of the changes. We won't actually merge this until it has finished the development process.

tghosth commented:
Note some original comments were raised here:
https://github.com/OWASP/ASVS/pull/1494#pullrequestreview-1622483377

tghosth commented on 5.0/en/0x51-V51-OAuth2.md:
```diff
@@ -2,111 +2,86 @@
-| **3.8.10** | [ADDED] Verify that Clients are utilizing the parameters "scope" and "authorization_details" to determine the requested/related resources and/or actions the access token are restricted to. | ✓ | ✓ | ✓ |
+| **51.2.1** | [ADDED] Verify that when an OAuth Client can interact with more than one Authorization Server, Clients should verify that the issuer "iss" parameter value is what it expected from the authorization response to prevent against mix-up attacks. [reference link for Mix-up attacks: https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics-18#mix_up] In the absence of "iss"  ...
+| **51.2.2** | [ADDED] Verify that the Client is using the PKCE flow or alternatively the OpenID Connect "nonce" parameter and the respective Claim in the ID Token. The PKCE challenge or OpenID Connect "nonce" must be transaction-specific and securely bound to the client and the user agent in which the transaction was started.  | ✓ | ✓ | ✓ |
+| **51.2.3** | [ADDED] Verify that if a Client sends a valid PKCE "code_challenge" parameter in the authorization request, the Authorization Server enforces the correct usage of "code_verifier" at the token endpoint. | ✓ | ✓ | ✓ |
```
@csfreak92 The way this is worded makes it sound like there requirement falls on the AS developer, not the Client developer so this requirement would be in the wrong section.

csfreak92 replied:
Woops, you are correct now that I read it a dozen times. Let me clean this up and push a commit soon.

tghosth replied:
Fixed in #1880

tghosth commented on 5.0/en/0x51-V51-OAuth2.md:
```diff
@@ -2,111 +2,86 @@
-
+| **51.4.1** | [ADDED] Verify that the Resource Owner password credentials grant is not used. This grant type insecurely exposes the credentials of the Resource Owner to the client, increasing the attack surface of the application.  | ✓ | ✓ | ✓ |
 
 ## Mapping
```
Is this just for our internal use to see how the new sections map to the old sections @csfreak92 ?

csfreak92 replied:
Hi @tghosth, yes we can use this for mapping internally if we want to or if we would decide to publish it as a part of helping the users see which requirements map to each subsection. Your call if you wish me to add it for the official doc or keep this internally. If for internal mapping usage, where can I place this?

tghosth replied:
Fine to leave it here for now and remove it at the end

csfreak92 replied:
You mean will remove it on the final commit, yes?

tghosth commented on 5.0/en/0x51-V51-OAuth2.md:
```diff
@@ -0,0 +1,99 @@
+* OAuth 2.0 Best Practices: https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics#name-best-practices
+* Mix-up attacks: https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics-18#mix_up
+* RFC9207 [I-D.ietf-oauth-iss-auth-resp]: https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics-18#section-2.1-4
+* & [I-D.ietf-oauth-iss-auth-resp]: https://datatracker.ietf.org/doc/html/draft-ietf-oauth-iss-auth-resp-00
```
These two lines look a little confusing. What are the supposed to be pointing to? The same thing or different things? @csfreak92

csfreak92 replied:
Sorry, which ones please @tghosth?

tghosth replied:
> * RFC9207 [I-D.ietf-oauth-iss-auth-resp]: https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics-18#section-2.1-4
> * & [I-D.ietf-oauth-iss-auth-resp]: https://datatracker.ietf.org/doc/html/draft-ietf-oauth-iss-auth-resp-00

These two lines, I can't understand what they are supposed to be pointing to or telling me.

csfreak92 replied:
They are pointing to different things. One is for OAuth 2.0 Security Best Current Practice and the other one is for OAuth 2.0 Authorization Server Issuer Identifier in Authorization Response. I think the confusion was with the titles in the [hard brackets here]. Let me fix that too.

tghosth replied:
Fixed in #1880

elarlang commented:
51.4.1 - We can not have requirements for the Resource Owner (end-user). Is certain flow or grant accepted/supported, is the Authorization Server responsible for configuring it for the OAuth Client? So it belongs to the "Authorization Server" section.

This is for a quick fix. But... as I opened a discussion in https://github.com/OWASP/ASVS/issues/996#issuecomment-1858764702, maybe we should not list the grants/flows which are not allowed per requirement, but have one requirement to say "Use the PKCE flow and X only, which also means, y z w flows are not allowed."

csfreak92 commented on 5.0/en/0x51-V51-OAuth2.md:
```diff
@@ -0,0 +1,99 @@
+
+There are various different personas in the OAuth process, described in more detail in the terminology section below. The requirements in this chapter are structured based on those personas as requirements for one persona may not be relevant for a different persona.
+
+## Terminology
```
Sure, I can move them to the bottom @tghosth. Sorry for the late response.

tghosth replied:
Fixed in #1880

csfreak92 commented on 5.0/en/0x51-V51-OAuth2.md:
```diff
@@ -0,0 +1,99 @@
+* Resource Owner (RO) - refers to an entity capable of granting access to a protected resource. When the resource owner is a person, it is referred to as an end-user.
+* Resource Server (RS) - refers to the server hosting the protected resources, capable of accepting and responding to protected resource requests using access tokens.
+
+## OAuth 2.0 Basics
```
Hi @tghosth, do you mean the headers such as ## OAuth 2.0 Basics, ## PKCE, and so forth? It was to highlight some of the risks and basic info about each of those subsections for implementing OAuth and to provide some useful info such as what is PKCE for, what we are preventing with resource owner password credentials grant, etc. I guess you were correct it is to set out the key risks and provide some additional info for each usage.

tghosth replied:
Yeah, I think we need to either:

1. Eliminate this text
2. Move the text into the glossary if it is explanatory
3. Move the text into the section it refers to if it is important context for implementing or verifying the requirements.

csfreak92 replied:
> 2. Move the text into the glossary if it is explanatory
> 3. Move the text into the section it refers to if it is important context for implementing or verifying the requirements.

@tghosth, option 3 doesn't work as we have chopped the requirements into implementer-based view (that's what the mapping section is for -- to keep track which requirement goes into which subsection. Best I can do is just move this whole block of explanatory text/heading into the glossary area right after the terminology subsection. 

@elarlang, I like the way the layout/format for the terminology from that link. I'm following that, thanks!

csfreak92 commented on 5.0/en/0x51-V51-OAuth2.md:
```diff
@@ -2,111 +2,86 @@
+| **51.1.1** | [ADDED] Verify that the Authorization Server supports CSRF protection either using the mechanism provided by PKCE, the "nonce" parameter or the "state" parameter. | ✓ | ✓ | ✓ |
+| **51.1.2** | [ADDED] Verify that the replay of authorization codes into the authorization response is prevented either by using the PKCE flow or alternatively the OpenID Connect "nonce" parameter and the respective Claim in the ID Token. The PKCE challenge or OpenID Connect "nonce" must be transaction-specific and securely bound to the client and the user agent in which the transaction was star ...
+| **51.1.3** | [ADDED] Verify that Authorization Servers are mitigating PKCE Downgrade Attacks by ensuring a token request containing a "code_verifier" parameter is accepted only if a "code_challenge" parameter was present in the authorization request. | ✓ | ✓ | ✓ |
+| **51.1.4** | [ADDED] Verify that refresh tokens are sender-constrained or use refresh token rotation. | ✓ | ✓ | ✓ |
```
Noted. Will add that detail as well.

tghosth replied:
Fixed in #1880

tghosth commented:
hi @csfreak92, are you still clear on the remaining feedback that needs to be actioned here. It would be great to get a first draft of this chapter merged into master.

csfreak92 commented:
Yes @tghosth, still working on it in my local copy first before pushing a commit. I will push a commit soon around late next week after this week's madness. :)

csfreak92 commented on 5.0/en/0x51-V51-OAuth2.md:
```diff
@@ -2,111 +2,86 @@
-| **3.8.9** | [ADDED] Verify that Clients are utilizing the parameters "scope" and "resource", respectively to determine the Resource Server they want to access. | ✓ | ✓ | ✓ |
-| **3.8.10** | [ADDED] Verify that Clients are utilizing the parameters "scope" and "authorization_details" to determine the requested/related resources and/or actions the access token are restricted to. | ✓ | ✓ | ✓ |
+| **51.2.1** | [ADDED] Verify that when an OAuth Client can interact with more than one Authorization Server, Clients should verify that the issuer "iss" parameter value is what it expected from the authorization response to prevent against mix-up attacks. [reference link for Mix-up attacks: https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics-18#mix_up] In the absence of "iss"  ...
+| **51.2.2** | [ADDED] Verify that the Client is using the PKCE flow or alternatively the OpenID Connect "nonce" parameter and the respective Claim in the ID Token. The PKCE challenge or OpenID Connect "nonce" must be transaction-specific and securely bound to the client and the user agent in which the transaction was started.  | ✓ | ✓ | ✓ |
```
Finalizing my review on this to ensure this nonce's importance/usage is not overlooked. Will push a commit soon. Thanks!

tghosth replied:
Yes because the server needs to create it, not the client. Do you need to add that guidance somewhere else or is that already covered @csfreak92 ?

tghosth replied:
I saw that you added this in the "PKCE" section. Do we not need to mention this somewhere in a requirement @csfreak92 or do we already do so?

csfreak92 replied:
> I saw that you added this in the "PKCE" section. Do we not need to mention this somewhere in a requirement @csfreak92 or do we already do so?

@tghosth, I moved it to the PKCE section to cover that part about the nonce. I think it does sit well there, don't you think?

csfreak92 replied:
> Yes because the server needs to create it, not the client. Do you need to add that guidance somewhere else or is that already covered @csfreak92 ?

This one is covered. I moved that nonce comment to the PKCE section. :)

tghosth commented:
@csfreak92 please do the following:

* Review my changes here to make sure you are comfortable with them: https://github.com/OWASP/ASVS/pull/1880/commits/4821e9086a122e2108a2b656f9836b1901da2bba
    * I removed 5.1.5 as per Elar's comment https://github.com/OWASP/ASVS/pull/1804#issuecomment-1858766621 which I agree with.
    * Removed the link in a requirement as it appears in the references at the end.
* As per Elar's comment here: https://github.com/OWASP/ASVS/pull/1804#issuecomment-185876662 about 51.4.1 - We can not have requirements for the Resource Owner (end-user). Is a certain flow or grant is accepted/supported, it is the Authorization Server's responsibility to configure it for the OAuth Client. So it belongs to the "Authorization Server" section. - I agree, please open a PR to address this.
* Please answer my question here: https://github.com/OWASP/ASVS/pull/1804#discussion_r1514575536

jmanico commented:
I also have one suggestion:

For...

+| 51.3.1 | [ADDED] Verify that Resource Servers are using mechanisms for sender-constraining access tokens to prevent token replay, such as Mutual TLS for OAuth 2.0 or OAuth Demonstration of Proof of Possession (DPoP). | ✓ | ✓ | ✓ |

... you are supposed to use the same client secret given to the client during registration. This implied it's arbitrary. My edit suggestion is:

+| 51.3.1 | [ADDED] "Verify Resource Servers implement sender-constrained access token mechanisms to mitigate token replay risks. This is achieved by mandating Mutual TLS for OAuth 2.0 or OAuth DPoP, using the client secret provided at client registration.". | ✓ | ✓ | ✓ |

csfreak92 commented:
> ... you are supposed to use the same client secret given to the client during registration. This implied it's arbitrary. My edit suggestion is:
> 
> +| **51.3.1** | [ADDED] "Verify Resource Servers implement sender-constrained access token mechanisms to mitigate token replay risks. This is achieved by mandating Mutual TLS for OAuth 2.0 or OAuth DPoP, using the client secret provided at client registration.". | ✓ | ✓ | ✓ |

Noted, @jmanico. Just as I mentioned in the other comment, I thought of explicitly calling it out as some people might not think it is implied already, but for brevity I agree with this assessment :)

csfreak92 commented:
> This is for a quick fix. But... as I opened a discussion in [#996 (comment)](https://github.com/OWASP/ASVS/issues/996#issuecomment-1858764702), maybe we should not list the grants/flows which are not allowed per requirement, but have one requirement to say "Use the PKCE flow and X only, which also means, y z w flows are not allowed."

Oh sorry Elar, I thought I addressed this already. I will fix it in my new PR that I was asked to open.

tghosth commented:
@elarlang @jmanico let's get this approved and then we can see what other comments we have on it

elarlang reviewed (approved):
Let's get it merged to have it in the spotlight to get more attention.
